### DISCUSSION TITLE
Inconsistent SQLAlchemy logging with hidden timestamp type coercion

### DISCUSSION THREAD

**Discussion by maramizo** [sqlalchemy/sqlalchemy#11535] [Usage Questions] [answered]
### Describe the bug

Datetime is automatically coerced into `::timestamp` when passed to `func.timezone`, which causes unexpected behavior. This is not shown through SQLAlchemy logs and this coercion is completely hidden until actually viewing database logs.

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/core/engines.html#configuring-logging

### SQLAlchemy Version in Use

2.0.30

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL 13.14

### Python Version

3.10.11

### Operating system

Windows

### To Reproduce

```python
>> current_dt = func.timezone('US/Eastern', start_date)
>> current_time = func.cast(current_dt, TIME)
>> sel = select(current_time)
>> res = session.execute(sel)

>> print(res.fetchall())
[(datetime.time(19, 44),)]

# Engine logs
SELECT CAST(timezone(%(timezone_2)s, %(timezone_3)s) AS TIME WITHOUT TIME ZONE) AS timezone_1
[generated in 0.00014s] {'timezone_2': 'US/Eastern', 'timezone_3': datetime.datetime(2024, 6, 24, 15, 44)}

# Expected result based on engine logs
11:44:00

# Actual query to run
SELECT CAST(timezone('US/Eastern', '2024-06-24T15:44:00'::timestamp) AS TIME WITHOUT TIME ZONE) AS timezone_1
```

As shown above, logging actually shows a different query than the one that runs on the database. In this specific use case, this causes timezone to incorrectly add time when converting UTC -> US/Eastern instead of subtracting time, while the coercion happens silently and is not implied anywhere throughout sqlalchemy logs.

I imagine this is a base use-case for the timezone functionality and being able to properly parse in & display how datetimes are interacted with is a significant part of it.

### Error

N/A

**Comment by zzzeek** [comment, marked as answer]
hi -

if SQLAlchemy logging shows this:

    SELECT CAST(timezone(%(timezone_2)s, %(timezone_3)s) AS TIME WITHOUT TIME ZONE) AS timezone_1
    [generated in 0.00014s] {'timezone_2': 'US/Eastern', 'timezone_3': datetime.datetime(2024, 6, 24, 15, 44)}

then that's what SQLAlchemy is sending.   If you then see in your server logs this:

    SELECT CAST(timezone('US/Eastern', '2024-06-24T15:44:00'::timestamp) AS TIME WITHOUT TIME ZONE) AS timezone_1

psycopg2 does that, unconditionally when it sees a python datetime object.  if you dont want it to do that, then you'd need to not send a datetime() object into it, but more correctly it seems you probably want to cast around your datetime itself to be whatever kind of object you need.

it's not clear what the actual SQL you want emitted is, is it this?

    SELECT CAST(timezone('US/Eastern', '2024-06-24T15:44:00') AS TIME WITHOUT TIME ZONE) AS timezone_1

That would means you would send the datetime as a formatted string.  if you wanted this:

    SELECT CAST(timezone('US/Eastern', '2024-06-24T15:44:00'::timestamp without time zone) AS TIME WITHOUT TIME ZONE) AS timezone_1

then...im not sure what psycopg2 does, if you send a tz-naive datetime, you might want to just cast like that directly.

overall though the example shows SQLAlchemy sending through what you are telling it so there's no bug

**Reply by maramizo** [reply]
Thanks for the quick response. Do you know of any documentation on psycopg2 behaviors such as that one? This one definitely blindsided me & I'd like to know if there are other behaviors like this that I need to keep an eye out for.

**Reply by zzzeek** [reply]
the basic idea is here: https://www.psycopg.org/docs/usage.html#adaptation-of-python-values-to-sql-types  though it doesnt get into telling you what it actually looks like

**Reply by CaselIT** [reply]
Not sure if psycopg 3 does something different here?
Since it uses server side binding it's sqlalchemy that does the casting, but I'm not sure if this case is correctly handed
